collections/quad.go

```diff
@@ -0,0 +1,425 @@
+package collections
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"cosmossdk.io/collections/codec"
+	"cosmossdk.io/schema"
+)
+
+// Quad defines a multipart key composed of four keys.
+type Quad[K1, K2, K3, K4 any] struct {
+	k1 *K1
+	k2 *K2
+	k3 *K3
+	k4 *K4
+}
+
+// Join4 instantiates a new Quad instance composed of the four provided keys, in order.
+func Join4[K1, K2, K3, K4 any](k1 K1, k2 K2, k3 K3, k4 K4) Quad[K1, K2, K3, K4] {
+	return Quad[K1, K2, K3, K4]{&k1, &k2, &k3, &k4}
+}
+
+// K1 returns the first part of the key. If nil, the zero value is returned.
+func (t Quad[K1, K2, K3, K4]) K1() (x K1) {
+	if t.k1 != nil {
+		return *t.k1
+	}
+	return x
```

Comment:
nit: I would recommend using some utility function like
```go
func pointerToValue[T any](p *T) (v T) {
     if p != nil {
          return *p
     }
    return v
}
```
Or use an already library having such functions (there are some, imo, dangerous dereferencing later in the code that could benefit from it too)